File: src/dart/test/invariants_test.dart
import 'dart:convert';

import 'package:ores_dnd/ores_dnd.dart';
import 'package:test/test.dart';

import 'helpers.dart';

void main() {
  test('xorshift64* reproduces the Rust reference values', () {
    final rng = Xorshift64(7);
    expect(
      [rng.nextU64(), rng.nextU64(), rng.nextU64()],
      [-3315863738710675794, -1322813130528676178, 1648209070578717474],
    );
    final b = Xorshift64(7);
    expect([b.below(100), b.below(100), b.below(100)], [7, 63, 78]);
    expect(
      jsonEncode(Fuzz.randomSequence(42, 10).map((i) => i.toJson()).toList()),
      jsonEncode(Fuzz.randomSequence(42, 10).map((i) => i.toJson()).toList()),
    );
  });

  test(
    'the Dart generator reproduces every Rust-generated fuzz trace input for input',
    () {
      final fuzz = corpus()
          .where(
            (c) =>
                c.declaration == 'DndSessionTrace' &&
                c.expectation == 'accepted' &&
                c.file.startsWith('fuzz-'),
          )
          .map(
            (c) =>
                decodeDeclaration('DndSessionTrace', c.json) as DndSessionTrace,
          )
          .toList();
      expect(fuzz.length, greaterThanOrEqualTo(40));
      for (final trace in fuzz) {
        final seed = int.parse(trace.id.substring('fuzz-'.length), radix: 16);
        final generated = Fuzz.randomSequence(
          seed,
          trace.inputs.length,
        ).map((i) => jsonEncode(i.toJson())).toList();
        expect(
          generated,
          trace.inputs.map((i) => jsonEncode(i.toJson())).toList(),
          reason: trace.id,
        );
      }
    },
  );

  test('session invariants hold over random sequences', () {
    final seeds = Xorshift64(0xc0ffee00);
    for (var run = 0; run < 1500; run++) {
      final seed = seeds.nextU64();
      final session = DndSession();
      var previous = DndSessionSnapshot.idle;
      final inputs = Fuzz.randomSequence(seed, 24);
      for (var step = 0; step < inputs.length; step++) {
        final input = inputs[step];
        final snapshot = session.apply(input);
        final ctx =
            'seed ${seed.toRadixString(16)} step $step ${input.kind.wire} -> $snapshot';
        expect(
          DndSessionSnapshot.fromJson(snapshot.toJson()),
          snapshot,
          reason: ctx,
        );
        if (previous.state.isTerminal &&
            input.kind != DndSessionInputKind.start) {
          expect(snapshot, previous, reason: ctx);
        }
        switch (snapshot.state) {
          case DndSessionState.idle:
            expect(
              snapshot.dragId == null &&
                  snapshot.targetId == null &&
                  snapshot.operation == null,
              isTrue,
              reason: ctx,
            );
            expect(session.envelope, isNull, reason: ctx);
          case DndSessionState.dragging:
            expect(
              snapshot.dragId != null && snapshot.operation == null,
              isTrue,
              reason: ctx,
            );
            expect(
              snapshot.targetId != null,
              snapshot.errorCode != null,
              reason: ctx,
            );
          case DndSessionState.overTarget:
            expect(
              session.envelope!.allowedOperations,
              contains(snapshot.operation),
              reason: ctx,
            );
            expect(
              snapshot.targetId != null && snapshot.errorCode == null,
              isTrue,
              reason: ctx,
            );
            if (input.kind == DndSessionInputKind.enter &&
                snapshot != previous) {
              expect(
                input.policy!.allowedOperations,
                contains(snapshot.operation),
                reason: ctx,
              );
              expect(snapshot.targetId, input.policy!.targetId, reason: ctx);
            }
          case DndSessionState.dropped:
            expect(
              snapshot.operation != null &&
                  snapshot.targetId != null &&
                  snapshot.errorCode == null,
              isTrue,
              reason: ctx,
            );
            expect(session.result!.accepted, isTrue, reason: ctx);
            if (snapshot != previous) {
              expect(input.kind, DndSessionInputKind.drop, reason: ctx);
              expect(previous.state, DndSessionState.overTarget, reason: ctx);
              expect(previous.targetId, snapshot.targetId, reason: ctx);
              expect(input.targetId, snapshot.targetId, reason: ctx);
            }
          case DndSessionState.cancelled:
            expect(
              snapshot.errorCode != null && snapshot.operation == null,
              isTrue,
              reason: ctx,
            );
            expect(session.result!.accepted, isFalse, reason: ctx);
            expect(
              session.result!.errorCode,
              snapshot.errorCode!.wire,
              reason: ctx,
            );
        }
        if (input.kind != DndSessionInputKind.start &&
            !previous.state.isTerminal &&
            previous.state != DndSessionState.idle) {
          expect(snapshot.dragId, previous.dragId, reason: ctx);
        }
        previous = snapshot;
      }
    }
  });
}

File: src/ts/test/invariants.test.mjs
import test from "node:test";
import assert from "node:assert/strict";
import { DndSession, isTerminal } from "../dist/session.js";
import { Xorshift64, randomSequence } from "../dist/fuzz.js";
import { decodeSessionSnapshot } from "../dist/corpus.js";

test("xorshift64* reproduces the Rust reference values", () => {
  const rng = new Xorshift64(7);
  assert.deepEqual([rng.nextU64(), rng.nextU64(), rng.nextU64()], [0xd1fbaf7f728d2eaen, 0xeda46c77629da6aen, 0x16df9d6ac76bd322n]);
  const b = new Xorshift64(7);
  assert.deepEqual([b.below(100), b.below(100), b.below(100)], [7, 63, 78]);
  assert.deepEqual(randomSequence(42, 10), randomSequence(42, 10));
  assert.notDeepEqual(randomSequence(42, 10), randomSequence(43, 10));
});

test("session invariants hold over random sequences", () => {
  const seeds = new Xorshift64(0xc0ffee00);
  for (let run = 0; run < 1500; run += 1) {
    const seed = seeds.nextU64();
    const session = new DndSession();
    let previous = { state: "idle" };
    randomSequence(seed, 24).forEach((input, step) => {
      const snapshot = session.apply(input);
      const ctx = `seed ${seed.toString(16)} step ${step} ${input.kind} -> ${JSON.stringify(snapshot)}`;
      // 1. structurally valid and JSON round-trippable
      assert.deepEqual(decodeSessionSnapshot(JSON.parse(JSON.stringify(snapshot))), snapshot, ctx);
      // 2. terminal states absorb everything but start
      if (isTerminal(previous.state) && input.kind !== "start") assert.deepEqual(snapshot, previous, ctx);
      switch (snapshot.state) {
        case "idle":
          assert.ok(snapshot.dragId === undefined && snapshot.targetId === undefined && snapshot.operation === undefined, ctx);
          assert.equal(session.envelope, null, ctx);
          break;
        case "dragging":
          assert.ok(snapshot.dragId !== undefined && snapshot.operation === undefined, ctx);
          assert.equal(snapshot.targetId !== undefined, snapshot.errorCode !== undefined, ctx); // 3. rejecting target ⇔ reason
          break;
        case "over-target":
          assert.ok(session.envelope.allowedOperations.includes(snapshot.operation), ctx); // 4. source-allowed operation
          assert.ok(snapshot.targetId !== undefined && snapshot.errorCode === undefined, ctx);
          if (input.kind === "enter" && JSON.stringify(snapshot) !== JSON.stringify(previous)) {
            assert.ok(input.policy.allowedOperations.includes(snapshot.operation), ctx);
            assert.equal(snapshot.targetId, input.policy.targetId, ctx);
          }
          break;
        case "dropped":
          assert.ok(snapshot.operation !== undefined && snapshot.targetId !== undefined && snapshot.errorCode === undefined, ctx);
          assert.equal(session.result.accepted, true, ctx);
          if (JSON.stringify(snapshot) !== JSON.stringify(previous)) { // 5. drop lands only on the accepting target
            assert.equal(input.kind, "drop", ctx);
            assert.equal(previous.state, "over-target", ctx);
            assert.equal(previous.targetId, snapshot.targetId, ctx);
            assert.equal(input.targetId, snapshot.targetId, ctx);
          }
          break;
        case "cancelled":
          assert.ok(snapshot.errorCode !== undefined && snapshot.operation === undefined, ctx);
          assert.equal(session.result.accepted, false, ctx);
          assert.equal(session.result.errorCode, snapshot.errorCode, ctx);
          break;
        default:
          assert.fail(ctx);
      }
      // 6. dragId only changes on start
      if (input.kind !== "start" && !isTerminal(previous.state) && previous.state !== "idle") assert.equal(snapshot.dragId, previous.dragId, ctx);
      previous = snapshot;
    });
  }
});

test("the TypeScript generator reproduces every Rust-generated fuzz trace input for input", async () => {
  const { traces } = await import("./helpers.mjs");
  const fuzz = (await traces()).filter((t) => t.id.startsWith("fuzz-"));
  assert.ok(fuzz.length >= 40, `expected the fuzz corpus, got ${fuzz.length}`);
  for (const trace of fuzz) {
    const seed = BigInt(`0x${trace.id.slice("fuzz-".length)}`);
    assert.deepEqual(randomSequence(seed, trace.inputs.length), trace.inputs, trace.id);
  }
});
